File: app/src/main/java/com/example/mediaplayer/playlist/PlayListFragment.java
package com.example.mediaplayer.playlist;
import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import com.example.mediaplayer.R;
import com.example.mediaplayer.databinding.PlaylistFragmentBinding;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import static com.example.mediaplayer.constants.CHOSEN_SONG_INDEX;
import static com.example.mediaplayer.constants.LIST_SONG;


/**
 * A simple {@link Fragment} subclass.
 */
public class PlayListFragment extends Fragment {
    private NavController navController;
    private static final int MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE= 5;
    private PlaylistFragmentBinding binding;
    public PlayListFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
         binding=PlaylistFragmentBinding.inflate(inflater);
        //find the nav controller so i can use it to navigate
         navController=Navigation.findNavController(Objects.requireNonNull(getActivity()),R.id.nav_host_fragment);


        return binding.getRoot();
    }

    private void navigate(List<PlayListModel> playListModels, int itemClickedIndex)
    {
        Bundle bundle=new Bundle();
        bundle.putParcelableArrayList(LIST_SONG,(ArrayList)playListModels);
        bundle.putInt(CHOSEN_SONG_INDEX,itemClickedIndex);
        navController.navigate(R.id.action_playListFragment_to_chosenSongFragment,bundle);
    }

    @Override
    public void onStart() {
        super.onStart();
        checkpermission();

    }

    private void prepareMusicList()
    {
        List<PlayListModel>playListModels;
        PlayListViewModel playListViewModel = ViewModelProviders.of(Objects.requireNonNull(getActivity())).get(PlayListViewModel.class);
        playListModels=playListViewModel.getPlayLists();
        //creating adapter and set it with the playlists
        playlistAdapter adapter= new playlistAdapter(playListModels, this::navigate);
        //setup recyclerview with adapter
        binding.playList.setAdapter(adapter);
        if(playListModels==null)
        {
            binding.noAudioText.setVisibility(View.VISIBLE);
        }
    }

    private void checkpermission()
    {
        // Here, thisActivity is the current activity
        if (ContextCompat.checkSelfPermission(Objects.requireNonNull(getActivity()),
                Manifest.permission.READ_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            // Permission is not granted
            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
                    Manifest.permission.READ_EXTERNAL_STORAGE)) {
                Toast.makeText(getActivity(),"You should grant this permission so application can access your audio files",Toast.LENGTH_LONG).show();
                // No explanation needed; request the permission
                requestPermissions(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},
                        MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE);

            } else {
                // No explanation needed; request the permission
                requestPermissions(
                        new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},
                        MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE);

            }
        } else {

            // Permission has already been granted
           prepareMusicList();

        }
    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode,permissions,grantResults);
        if (requestCode == MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE) {
            // If request is cancelled, the result arrays are empty.
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // Permission has already been granted
               prepareMusicList();
            }
        }

    }


}

File: app/src/main/java/com/example/mediaplayer/ui/playlist/PlayListFragment.kt
/*
 * Copyright 2019 Abdelrhman Sror. All rights reserved.
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *   http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package com.example.mediaplayer.ui.playlist

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.example.mediaplayer.R
import com.example.mediaplayer.database.provideDatabase
import com.example.mediaplayer.databinding.PlaylistFragmentBinding
import com.example.mediaplayer.intent.CHOSEN_SONG_INDEX
import com.example.mediaplayer.model.SongModel
import com.example.mediaplayer.model.toSongModel
import com.example.mediaplayer.repositry.provideTrackRepository
import com.example.mediaplayer.ui.ClickType
import com.example.mediaplayer.ui.OnItemClickListener
import com.example.mediaplayer.viewModels.PlayListViewModel
import com.example.mediaplayer.viewModels.PlayListViewModelFactory
import com.example.mediaplayer.viewModels.createViewModel
import com.example.mediaplayer.viewModels.createViewModelFactory


/**
 * A simple [Fragment] subclass.
 */
class PlayListFragment : Fragment() {

    lateinit var viewModelFactory: PlayListViewModelFactory

    lateinit var viewModel: PlayListViewModel

    private lateinit var navController: NavController
    private lateinit var binding: PlaylistFragmentBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        binding = PlaylistFragmentBinding.inflate(inflater)
        viewModelFactory = createViewModelFactory(requireActivity().application)!!
        viewModelFactory.setData(provideTrackRepository(requireActivity().application, provideDatabase(requireActivity().application)))
        viewModel = createViewModel(this, viewModelFactory)
        //find the nav controller so i can use it to navigate
        navController = Navigation.findNavController(activity!!, R.id.nav_host_fragment)
        prepareMusicList()
        binding.viewmodel = viewModel
        binding.lifecycleOwner = this
        return binding.root
    }


    private fun prepareMusicList() {
        setUpPlayList()
        //observing any changes to the playlist
        viewModel.playLists.observe(viewLifecycleOwner, androidx.lifecycle.Observer { playlist ->
            if (playlist.isNullOrEmpty()) {
                showPlaylistEmptyText()
                // binding.bottomSheetLayout.visibility = View.GONE
            } else {
                hidePlaylistEmptyText()

            }
            submitPlaylistToAdapter(playlist.toSongModel())

        })

    }

    private fun submitPlaylistToAdapter(songModel: List<SongModel>) {
        (binding.listSong.adapter as PlaylistAdapter).submitList(songModel)

    }

    private fun hidePlaylistEmptyText() {
        binding.noAudioText.visibility = View.GONE
    }

    private fun showPlaylistEmptyText() {
        binding.noAudioText.visibility = View.VISIBLE
    }

    private fun setUpPlayList() {
        //setup recycler view with adapter
        binding.listSong.apply {
            this.adapter = createAdapter()
        }

    }

    private fun createAdapter(): PlaylistAdapter {
        //creating adapter and set it with the playlist
        return PlaylistAdapter(object : OnItemClickListener {
            override fun onClick(clickType: ClickType, itemClickIndex: Int) {
                this@PlayListFragment.onClick(clickType, itemClickIndex)
            }
        })
    }

    private fun onClick(clickType: ClickType, itemClickIndex: Int) {
        when (clickType) {
            ClickType.RUN -> navigateAndPlayAudio(itemClickIndex)
            ClickType.FAVOURITE -> addToFavourite(viewModel.playLists.value!![itemClickIndex].toSongModel())
            ClickType.EDIT -> {
            }
            ClickType.DELETE -> {
            }
        }
    }

    private fun addToFavourite(songModel: SongModel) {
        viewModel.addToFavourite(songModel)
    }

    private fun navigateAndPlayAudio(position: Int) {
        val args = Bundle()
        args.putInt(CHOSEN_SONG_INDEX, position)
        navController.navigate(R.id.action_playListFragment_to_chosenSong_dest, args)
    }


}// Required empty public constructor



